src/core/namespace.c

```diff
@@ -815,18 +815,24 @@ static int make_read_only(MountEntry *m, char **blacklist) {
         return r;
 }
 
-static bool namespace_info_mount_apivfs(const NameSpaceInfo *ns_info) {
+static bool namespace_info_mount_apivfs(const char *root_directory, const NameSpaceInfo *ns_info) {
         assert(ns_info);
 
-        /* ProtectControlGroups= and ProtectKernelTunables= imply MountAPIVFS=, since to protect the API VFS mounts,
-         * they need to be around in the first place... */
+        /*
+         * ProtectControlGroups= and ProtectKernelTunables= imply MountAPIVFS=,
+         * since to protect the API VFS mounts, they need to be around in the
+         * first place... and RootDirectory= or RootImage= need to be set.
+         */
 
-        return ns_info->mount_apivfs ||
-                ns_info->protect_control_groups ||
-                ns_info->protect_kernel_tunables;
+        /* root_directory should point to a mount point */
+        return root_directory &&
+                (ns_info->mount_apivfs ||
+                 ns_info->protect_control_groups ||
+                 ns_info->protect_kernel_tunables);
```

Comment:
> hmm, so, thinking about this again, what does this actually change except for making these conditoins a bit more complex?

> Note that before we actually mount the relevant file systems we check with path_is_mount_point() whether there already is a mount point in place. Also, we don't support systems that don't have /sys, /proc and /dev mounted on the host (or to say this differently: if PID 1 finds them not mounted it will mount them automatically, ensuring they are always there on the host). These two facts mean that MountAPIVFS has zero effect already when not combined with RootDirectory=: the dirs will already be mounted, and it will notice and mount again...

Hmm, still I can see that MountAPIVFS= has an effect, if we only set MountAPIVFS= without root directory nor root image, we end up creating a new private namespace, since it detects that we need a new mount namespace, and we remount it later as a slave. Of course we do not mount them again since they are already mounted however we are already in a new private mountns. Maybe I did not read the man pages correctly ? but I thought that if Root directory is not set MountAPIVFS= should have zero affect, I'm saying this since we had several issues reported by users where they were surprised when they end up in a private mountns where mounts do not propagate back... hence this change to activate MountAPIVFS= only when root directory is changed. 

These patches are also to make sure that MountAPIVFS= /proc , /sys and /dev will work under `RootDirecotry=` or `RootImage=,` if the paths do not exist. I'm not referring to PID 1 but other services under these conditions, currently MountAPIVFS= is not honored if the paths do not exist, the paths never show up under a different RootDirectory.  Also the `path_is_mount_point()` that you refer to of course will make it a nop under PID 1 however for other services with a different `RootDirectory=` if the path do not exist we fail, we never reach that code:
apply_mount() => mount_entry_chase() => chase_symlinks() https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L682

The `m->ignore` Boolean is set, we ignore that mount and we proceed, thus we never reach the `path_is_mount_point()` code. The /rootdirectory/{proc|sys|dev} are silently ignored if they do not exist and `MountAPIVFS=yes` is not honored. I tested this.

Now after your comments I just noticed also that this change is not perfect, since we may not calculate mount_apivfs mounts if RootDirectory= or RootImage= are not set, and we fail to set `make_slave` https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L944

Now if you think that MountAPIVFS= should not create a private mountns, then beside these patches we have also to check RootDirectory= and RootImage= when we check MountAPIVFS=  https://github.com/systemd/systemd/blob/master/src/core/execute.c#L1673

> Hence, is there something this patch really corrects?

This patch tries to prevent MountAPIVFS= from creating a new private mountns if root directory or RootImage= are not set.

All the patches try to ensure that MountAPIVFS= will work if the paths do not exist under Root directory. I created the following issue for it: https://github.com/systemd/systemd/issues/5488